src/sentry/utils/snuba.py

```diff
@@ -643,16 +631,60 @@ def __init__(
         self.kwargs = kwargs
 
 
-def raw_query(start, end, groupby=None, conditions=None, filter_keys=None,
+class DefaultRetentionQueryStrategy(object):
+    """
+    When querying Snuba, apply start and end times depending on the user's
+    project's retention settings
+    """
+
+    def __init__(self):
+        self.__current_time = timezone.now()
+
+    def execute(self, project_id):
+        end = timezone.now()
+
+        project = Project.objects.get(pk=project_id)
+        retention = quotas.get_event_retention(
+            organization=Organization(project.organization_id)
+        )
+        if retention:
+            start = max(self.__current_time, datetime.utcnow() - timedelta(days=retention))
+        else:
+            start = datetime.utcnow() - timedelta(days=90)
+
+        # TODO(manu): can just record times in utc and then we dont need to naiveify
+        return naiveify_datetime(start), naiveify_datetime(end)
+
+
+class CustomQueryStrategy(object):
+    """
+    Apply custom start and end times, without checking retention settings
+    """
+
+    def __init__(self, start, end):
+        self.__start = start
+        self.__end = end
+
+    def execute(self, project_id):
+        # TODO(manu): maybe also just check with default retention
+        # OR: create a new strategy that does check - can be used when
+        # we directly send params from the UI to Snuba.
+
+        # convert to naive UTC datetimes, as Snuba only deals in UTC
+        # and this avoids offset-naive and offset-aware issues
+        return naiveify_datetime(self.__start), naiveify_datetime(self.__end)
```

Comment:
Even if you wanted to skip retention (for this, see @markstory question). You still want to throw if the interval is invalid to preserve consistency.